bin/runtime-common/src/refund_relayer_extension.rs

```diff
@@ -0,0 +1,844 @@
+		+ UtilityConfig<RuntimeCall = CallOf<R>>
+		+ GrandpaConfig<GI>
+		+ ParachainsConfig<PI, BridgesGrandpaPalletInstance = GI>
+		+ MessagesConfig<MI>
+		+ RelayersConfig,
+	GI: 'static + Send + Sync,
+	PI: 'static + Send + Sync,
+	MI: 'static + Send + Sync,
+	BE: 'static
+		+ Send
+		+ Sync
+		+ Default
+		+ SignedExtension<AccountId = R::AccountId, Call = CallOf<R>>,
+	PID: 'static + Send + Sync + Get<u32>,
+	LID: 'static + Send + Sync + Get<LaneId>,
+	FEE: 'static + Send + Sync + TransactionFeeCalculation<<R as RelayersConfig>::Reward>,
+	<R as frame_system::Config>::RuntimeCall:
+		Dispatchable<Info = DispatchInfo, PostInfo = PostDispatchInfo>,
+	CallOf<R>: IsSubType<CallableCallFor<UtilityPallet<R>, R>>
+		+ IsSubType<CallableCallFor<GrandpaPallet<R, GI>, R>>
+		+ IsSubType<CallableCallFor<ParachainsPallet<R, PI>, R>>
+		+ IsSubType<CallableCallFor<MessagesPallet<R, MI>, R>>,
+	<R as GrandpaConfig<GI>>::BridgedChain:
+		Chain<BlockNumber = RelayBlockNumber, Hash = RelayBlockHash, Hasher = RelayBlockHasher>,
+	<R as MessagesConfig<MI>>::SourceHeaderChain: SourceHeaderChain<
+		MessagesProof = FromBridgedChainMessagesProof<HashOf<BridgedChain<R, GI>>>,
+	>,
+{
+	const IDENTIFIER: &'static str = "RefundRelayerForMessagesFromParachain";
+	type AccountId = R::AccountId;
+	type Call = CallOf<R>;
+	type AdditionalSigned = ();
+	type Pre = Option<PreDispatchData<R::AccountId>>;
+
+	fn additional_signed(&self) -> Result<(), TransactionValidityError> {
+		Ok(())
+	}
+
+	fn validate(
+		&self,
+		_who: &Self::AccountId,
+		_call: &Self::Call,
+		_info: &DispatchInfoOf<Self::Call>,
+		_len: usize,
+	) -> TransactionValidity {
+		Ok(ValidTransaction::default())
+	}
+
+	fn pre_dispatch(
+		self,
+		who: &Self::AccountId,
+		call: &Self::Call,
+		post_info: &DispatchInfoOf<Self::Call>,
+		len: usize,
+	) -> Result<Self::Pre, TransactionValidityError> {
+		// reject batch transactions with obsolete headers
+		if let Some(UtilityCall::<R>::batch_all { ref calls }) = call.is_sub_type() {
+			for nested_call in calls {
+				let reject_obsolete_transactions = BE::default();
+				reject_obsolete_transactions.pre_dispatch(who, nested_call, post_info, len)?;
+			}
+		}
+
+		// now try to check if tx matches one of types we support
+		let parse_call_type = || {
+			if let Some(UtilityCall::<R>::batch_all { ref calls }) = call.is_sub_type() {
+				if calls.len() == 3 {
+					return Some(CallType::AllFinalityAndDelivery(
+						extract_expected_relay_chain_state::<R, GI>(&calls[0])?,
+						extract_expected_parachain_state::<R, GI, PI, PID>(&calls[1])?,
+						extract_messages_state::<R, GI, MI, LID>(&calls[2])?,
+					))
+				}
+				if calls.len() == 2 {
+					return Some(CallType::ParachainFinalityAndDelivery(
+						extract_expected_parachain_state::<R, GI, PI, PID>(&calls[0])?,
+						extract_messages_state::<R, GI, MI, LID>(&calls[1])?,
+					))
+				}
+				return None
+			}
+
+			Some(CallType::Delivery(extract_messages_state::<R, GI, MI, LID>(call)?))
+		};
```

Comment:
Can we have this logic defined on the `CallType` struct in a constructor ? Something like `from_batch_all_call()`. Also could we define `extract_expected_relay_chain_state` on `ExpectedRelayChainState` as `from_call()` or `from_grandpa_call()` ? `extract_expected_parachain_state` on `ExpectedParachainState`, `extract_messages_state`  on `MessagesState` ?

Reply:
That's a good idea, but then I have to duplicate this all 30-line `where` in this `from_batch_all_call`, which I don't like :) Perhaps get back to this when we'll make it a bit easier to configure (decrease `where` LOCs)?

Reply:
Sure, sounds good.

Reply:
Added note to #1666